plugins/woocommerce-blocks/assets/js/base/components/combobox/index.tsx

```diff
@@ -88,72 +147,193 @@ const Combobox = ( {
 		setValidationErrors,
 	] );
 
-	// @todo Remove patch for ComboboxControl once https://github.com/WordPress/gutenberg/pull/33928 is released
-	// Also see https://github.com/WordPress/gutenberg/pull/34090
+	const outerWrapperClasses = classnames(
+		'wc-block-components-combobox',
+		className || '',
+		{
+			'is-active': selectedOption?.value,
+			'has-error': error?.message && ! error?.hidden,
+		}
+	);
+
+	const ariaInvalid = error?.message && ! error?.hidden ? 'true' : 'false';
+	const activeValue = store.useState( 'activeValue' );
+	const [ inputFocussed, setInputFocussed ] = useState( false );
+
+	const onClose = useCallback( () => {
+		// If the search term doesn't match the selected option, try do an exact value match.
+		// e.g. if the user leaves "NZ" in the search box, select "New Zealand" on close.
+		// If not just leave the last selected value.
+		if ( searchTerm && selectedOption?.label !== searchTerm ) {
+			const exactValueMatch = findExactMatchBy(
+				'value',
+				searchTerm,
+				options
+			);
+
+			if ( exactValueMatch ) {
+				store.setSelectedValue( exactValueMatch.value );
+			} else {
+				setSearchTerm( selectedOption?.label || '' );
+			}
+		} else {
+			setSearchTerm( selectedOption?.label || '' );
+		}
+	}, [ setSearchTerm, selectedOption, searchTerm, options, store ] );
+
+	const minHeight =
+		matchingSuggestions.length * 35 > 300
+			? 300
+			: matchingSuggestions.length * 35;
+
 	return (
 		<div
-			id={ controlId }
-			className={ classnames( 'wc-block-components-combobox', className, {
-				'is-active': value,
-				'has-error': error?.message && ! error?.hidden,
-			} ) }
 			ref={ controlRef }
+			id={ controlId }
+			className={ outerWrapperClasses }
 		>
-			<ComboboxControl
-				className={ 'wc-block-components-combobox-control' }
-				label={ label }
-				onChange={ onChange }
-				onFilterValueChange={ ( filterValue: string ) => {
-					if ( filterValue.length ) {
-						// If we have a value and the combobox is not focussed, this could be from browser autofill.
-						const activeElement = isObject( controlRef.current )
-							? controlRef.current.ownerDocument.activeElement
-							: undefined;
-
-						if (
-							activeElement &&
-							isObject( controlRef.current ) &&
-							controlRef.current.contains( activeElement )
-						) {
-							return;
-						}
+			<div className="components-combobox-control components-base-control wc-block-components-combobox-control">
+				<ComboboxProvider
+					store={ store }
+					value={ searchTerm }
+					selectedValue={ selectedOption?.value || '' }
+					setValue={ ( val: string ) => {
+						startTransition( () => {
+							setSearchTerm( val );
 
-						// Try to match.
-						const normalizedFilterValue =
-							filterValue.toLocaleUpperCase();
+							// If the input is not focussed, it's autofill, attempt the best match.
+							if ( ! inputFocussed ) {
+								const bestMatch = findBestMatchByLabel(
```

Comment:
This especially handles situations like an autofill of "United States" which matches `um` and `us`.